### ISSUE TITLE
[DataTable] How to extract localize.uasset from ucas files UE5.2?!?!

### ISSUE THREAD

**Issue by muslimcyber26** [SolicenTEAM/UEExtractor#10]
Hey, can you help me for extracting the uasset to csv from game Life is Strange Double Exposure?

I don't know to use it.

**Comment by JunkBeat** [comment]
Hi. First you need to find the mapping file for your game (usmap) on the internet. Then extract the .uasset files using Fmodel. To generate the csv run the command: `UEExtractor.exe "...\FmodelOutput\Exports\YourGame" extracted_text.csv --noparallel`

**Comment by muslimcyber26** [comment]
Hey, I used that command. But, the result is none of the uasset extracted... After generated the csv, the csv files is not contain any single text from uasset..

**Comment by JunkBeat** [comment]
This is weird. 🤔 Can you upload sample files to Google Drive? For example, find some .uasset by text content via TotalCommander (I hope this will work).

**Comment by muslimcyber26** [comment]
Okey. I I'll uploaded to here some files for sample and all-uasset localize in other hosting.

This some sample that cannot be extracted by other tools (that have the same function)
[LiSDE.Some.Sample.zip](https://github.com/user-attachments/files/18178242/LiSDE.Some.Sample.zip)

And this All-Text can't read with your tools: [LiSDE-All-Text](https://workupload.com/file/S5LPjr7uhgq)

**Comment by DenisSolicen** [comment]
Well, it was fast, the problem is really in the structure of the `uasset` file, it is expected that the text structure starts with hex **`0x29 0x01 0x1F`** which does not happen, since the text starts with hex `0x00 0x00 0x00`. 

### For example from `EP1_Global_Localization.uasset` here:
`
00 00 00 49 27 6D 20 6E 6F 74 20 61 62 6F 76 65 20 72 65 6C 61 78 69 6E 67 20 6D 79 20 70 72 69 6E 63 69 70 6C 65 73 20 74 6F 20 67 69 76 65 20 53 61 66 69 20 74 68 61 74 20 62 6F 74 74 6C 65 20 63 61 70 2E 00
`
### Output:
- `
I'm not above relaxing my principles to give Safi that bottle cap.
`

And no, replacing the structure with `0x00 0x00 0x00` does not save the situation and does not give that data.
### And so this is the intermediate result:


The problem is deeper somewhere, and I will continue to dig, and try to adapt the program to such a structure, although this will give a huge amount of false (broken) data in the future.

**Comment by DenisSolicen** [comment]
@muslimcyber26 and, for double check, I currently download **Life is Strange Double Exposure** to accurately verify that this is a problem with the structure, and not with the feature of packing the game with Denuvo or something else.

If I manage to extract the text, I will provide instructions below, otherwise I will inform you that it is not possible to do this with my program yet.

**Comment by muslimcyber26** [comment]
Hello, any problem has been solved? I still interested for your tools,.

Btw, happy holiday and new years.

**Comment by DenisSolicen** [comment]
@muslimcyber26 unfortunately I can't say anything good, I received several messages with similar problems in UE5 (and only in UE5) and after looking at several structures of different files, I found that in all of them it is similar, but I also found many differences between these assets, and this is a problem, a BIG problem.

Now **Saipan** is trying to find a way to solve this problem, since he himself needs the work of this program, but I can’t say anything yet, the structure is radically different from UE4, in almost everything.

For Zen Loader files, a different approach is needed and completely written from scratch as new functionality, but for this I need a working `locres` file that contains the correct new structure of hashes and values, since now if we rely on the old one, nothing will work, since the offsets and even hashes are not equal to 32 characters (which was a constant for a hash) so I, of course, am trying to find a solution, but the likelihood of this is not high yet.

 Happy holiday and Happy New Year, you too.

**Comment by muslimcyber26** [comment]
Actually, I have tried it with Amrsheen's Localization Tools (Method 2), but as a result not all Uasset files already extracted from FMODEL can be extracted through Amrsheen tools. You can see the rest of the files that I couldn't extract with Amrsheen tools in “LiSDE.Some.Sample.zip”... And the rest of the uasset can be repacked with “UnrealReZen” from Noobincoding, it does work well in-game test.

Look this preview, if I using with Amrsheen Tools:
![LISDE1_result](https://github.com/user-attachments/assets/b7e56dac-cbe6-483f-8e96-eb4e1d841581)


But when I tried to extract the .pak. utoc. ucas file via ZenTools (45 GB uncooked, after extracted). Indeed, the results of the “localization” file were all extracted. And not just uasset, but uexp and uasset. Now, the problem is. Even though I use these two tools, if I use your Tools, it still can't be extracted into 'extracted_text.csv', and I tried to making repacking to ucas with UnrealReZen, just uasset or uexp only can repack. Not both of them (uasset and uexp), the result it's 'GAME CRASHED'

In case, you want extracted files from ZenTools (uexp&uasset) localization files. I can send them, for you to check that the tools work properly.

This extracted with ZenTools: [ZenTools.Extracted.EP1.ONLY.zip](https://github.com/user-attachments/files/18307631/ZenTools.Extracted.EP1.ONLY.zip)

I can waiting for your hardworking tools. Cause I can't start the project yet.

**Comment by DenisSolicen** [comment]
@muslimcyber26 Wow! It was very useful, your files gave a complete understanding of what the problem was, and I became very close to solving it, it remains to figure out how to compare the data from these two files, since I do not yet understand where the keys went. 

I can extract the text right now, but without the keys, it will create a `.locres` file, which means it will remain as a CSV file and will require manual repacking.

![image](https://github.com/user-attachments/assets/4d7d237c-a034-49ca-af06-24e12074afa4)

I also thought at first that the keys are now those lines in capital letters, but sometimes they are not enough for all the data that is in the files, which is somehow confusing.

## One more thing about UE5
I was learning about the features of UE5, and I found out that now game files can be both **compressed by Oodle** and **encrypted** (even if it seems that this is not the case), especially if any security system is used, and especially **Denuvo** or EAC.

**Comment by DenisSolicen** [comment]
@muslimcyber26 Apologies for the wait, I'm still dealing with this issue. As you understand, extracting the data is not a problem, the problem is extracting it correctly for the locres file.

## A little explanation:
So far, I can only say that according to all the tests conducted by me and Saipan, we can conclude that all UE versions up to and including 5.0 are fully supported, followed by the possibility of support and partial support, since the file structure can be either old (up to version 5.1 or new 5.2 and higher) and this It provides at least partial support for some projects, but I can't say that in your case.

## DataTable:
Plus, in the files that you provided, the text has a DataTable format, which is of course great for developers when creating localization, however...as it were, it is completely unsuitable for `locres` files. 

Although I may be wrong, and maybe you have an example of extracting a DataTable into a form suitable for `locres`, since even after solving the problem with keys and their matching, I will not be able to extract the text from the Table correctly, `locres` is not designed for this.

## To summarize:
I am close to solving the problem with the inability to extract text (but incorrectly from the DataTable).

If you have an example of normal text extraction from this format from other programs, then I ask you to provide it for study, as I have not seen such cases. I apologize for the problems that have arisen with my program, and thank you for your patience in solving these problems, have a nice day.

**Comment by muslimcyber26** [comment]
Here are all the text files that can be extracted. Because I only use the tools “FMODEL, ZenTools, and UE4Localization Tools”.
This has been extracted to a normal text file, but only those that can be extracted.

All Text for Research: 
[LiSDE.All.Text.with2Tools.zip](https://github.com/user-attachments/files/18341984/LiSDE.All.Text.with2Tools.zip)

I hope you can research it. For the DataTable problem, I don't understand. Because I have searched through FMODEL, but I didn't find it. Sorry about that, btw, thanks for helping.

**Comment by DenisSolicen** [comment]
I'm closing this issue, because so far I don't even have the ability and knowledge to manage the these strings through `.locres`, its whole essence, only confuses me with its existence in this repository. I can't operate with these lines through `.locres`, and don't even tell me that others can, they can extract, but not insert to working `locres`, or I don't know something. 

Sorry for the long wait, but my programs won't help you solve your problems, have a nice day.

**Comment by muslimcyber26** [comment]
It's solved right now. I'm using the WRONG METHOD.

Finally it works.
